### PR TITLE
Add reactor extensions.

### PR DESCRIPTION
**Pull request by utwyko**
Thanks for creating this library, it solves a problem I have constantly while working with Reactive Streams.

I've added a module with extensions for the implementation I work with the most often: [Project Reactor](https://projectreactor.io/).

### CODE REVIEW

File: koptional-reactor-extensions/build.gradle [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+apply from: '../gradle/publishing.gradle'
+
+dependencies {
+    compile libraries.koptional
+    compile libraries.reactor
+}
+
+dependencies {
+    testCompile libraries.spek
+    testCompile libraries.spekJunitPlatformEngine
```
Comment (artem-zinnatullin) [inline koptional-reactor-extensions/build.gradle]: Please add `testCompile libraries.junitPlatformLauncher` as you did in the previous PR #8  :)

File: koptional-reactor-extensions/src/main/kotlin/com/gojuno/koptional/reactor/reactor.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * Filters items emitted by a Flux by only emitting those that are `Some`.
+ * <p>
+ * <img width="640" height="310" src="https://raw.github.com/wiki/ReactiveX/RxJava/images/rx-operators/filter.png" alt="">
+ * <dl>
+ *  <dt><b>Scheduler:</b></dt>
+ *  <dd>{@code filter} does not operate by default on a particular {@link Scheduler}.</dd>
+ * </dl>
+ *
+ * @return a Flux that emits `Some.value` of those items emitted by the source Flux that are `Some`.
+ * @see <a href="http://reactivex.io/documentation/operators/filter.html">ReactiveX operators documentation: Filter</a>
```
Comment (artem-zinnatullin) [inline koptional-reactor-extensions/src/main/kotlin/com/gojuno/koptional/reactor/reactor.kt]: I guess we should use Reactor's pictures here https://raw.githubusercontent.com/reactor/reactor-core/v3.1.0.M2/src/docs/marble/filter.png

File: koptional-reactor-extensions/src/main/kotlin/com/gojuno/koptional/reactor/reactor.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * Filters items emitted by a Flux by only emitting those that are `None`.
+ * <p>
+ * <img width="640" height="310" src="https://raw.github.com/wiki/ReactiveX/RxJava/images/rx-operators/filter.png" alt="">
+ * <dl>
+ *  <dt><b>Scheduler:</b></dt>
+ *  <dd>{@code filter} does not operate by default on a particular {@link Scheduler}.</dd>
+ * </dl>
+ *
+ * @return a Flux that emits `Unit` for each item emitted by the source Flux that is `None`.
+ * @see <a href="http://reactivex.io/documentation/operators/filter.html">ReactiveX operators documentation: Filter</a>
```
Comment (artem-zinnatullin) [inline koptional-reactor-extensions/src/main/kotlin/com/gojuno/koptional/reactor/reactor.kt]: https://raw.githubusercontent.com/reactor/reactor-core/v3.1.0.M2/src/docs/marble/filter.png

File: koptional-reactor-extensions/src/main/kotlin/com/gojuno/koptional/reactor/reactor.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import reactor.core.publisher.Flux
+
+private inline fun <reified R : Any> Flux<*>.ofType(): Flux<R> = ofType(R::class.java)
+
+/**
+ * Filters items emitted by a Flux by only emitting those that are `Some`.
+ * <p>
+ * <img width="640" height="310" src="https://raw.github.com/wiki/ReactiveX/RxJava/images/rx-operators/filter.png" alt="">
+ * <dl>
+ *  <dt><b>Scheduler:</b></dt>
```
Comment (artem-zinnatullin) [inline koptional-reactor-extensions/src/main/kotlin/com/gojuno/koptional/reactor/reactor.kt]: Let's remove note about scheduling, it's standard for RxJava, but not part of javadoc for operators in Reactor

File: koptional-reactor-extensions/src/test/kotlin/com/gojuno/koptional/reactor/ReactorExtensionsSpec.kt [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    val stream by memoized {
+        Flux.just(Some("a"), None, Some("b"), Some("c"), None)
+    }
+
+    context("filterSome") {
+
+        val stepVerifier by memoized {
+            StepVerifier.create(
+                    stream.filterSome()
```
Comment (artem-zinnatullin) [inline koptional-reactor-extensions/src/test/kotlin/com/gojuno/koptional/reactor/ReactorExtensionsSpec.kt]: single line?

File: koptional-reactor-extensions/src/test/kotlin/com/gojuno/koptional/reactor/ReactorExtensionsSpec.kt [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+                    .expectNext("a", "b", "c")
+                    .verifyComplete()
+        }
+    }
+
+    context("filterNone") {
+
+        val subscriber by memoized {
+            StepVerifier.create(
+                    stream.filterNone()
```
Comment (artem-zinnatullin) [inline koptional-reactor-extensions/src/test/kotlin/com/gojuno/koptional/reactor/ReactorExtensionsSpec.kt]: single line?

**Comment by utwyko** [comment]
Thanks for the review, made changes accordingly.

**Comment by artem-zinnatullin** [comment]
Please give us a day or couple to discuss whether we'd like to make this part of Koptional repo or not since we're active RxJava users but not users of Project Reactor, I think decision will be positive but let's see :)

**Comment by artem-zinnatullin** [comment]
So we've discussed this and decided to merge, thanks for addition, will do release today!

**Comment by artem-zinnatullin** [comment]
@utwyko FYI we released 1.1.0 with your extensions, however Bintray did not link it to jcenter repo for some reason (as usual uh), I've contacted them and hope to fix this soon (last time similar issue took about 4 days to resolve).

For now your extension is available on https://dl.bintray.com/gojuno/maven/com/gojuno/koptional/, once issue will be resolved it should also become available on https://jcenter.bintray.com/com/gojuno/koptional/

**Comment by utwyko** [comment]
@artem-zinnatullin Great, thanks for the info.

**Comment by artem-zinnatullin** [comment]
Contacted Bintray https://stackoverflow.com/questions/44371758/bintray-does-not-link-new-artifact-to-jcenter

**Comment by artem-zinnatullin** [comment]
@utwyko Reactor extensions are now available on Jcenter

**Pull request merged**